lib/eventmesh/Etcd3EventMeshServer.js

```diff
@@ -0,0 +1,131 @@
+    const statusKey = `${resourceFolderName}/${CONST.RESOURCE_KEYS.STATE}`;
+    const lastOperationKey = `${resourceFolderName}/${CONST.RESOURCE_KEYS.LASTOPERATION}`;
+    return etcd.put(optionKey).value(val)
+      .then(() => etcd.put(statusKey).value(CONST.RESOURCE_STATE.IN_QUEUE))
+      .then(() => etcd.put(lastOperationKey).value(''));
+  }
+
+  updateResourceState(resourceType, resourceId, stateValue) {
+    return this.checkValidState(stateValue)
+      .then(() => this.updateResourceKey(resourceType, resourceId, CONST.RESOURCE_KEYS.STATE, stateValue));
+  }
+
+  updateResourceKey(resourceType, resourceId, key, value) {
+    const resourceFolderName = this.getResourceFolderName(resourceType, resourceId);
+    const resourceKey = `${resourceFolderName}/${key}`;
+    logger.debug(`Updating resource key: ${key} for ${resourceId}`);
+    return etcd.put(resourceKey).value(value);
+  }
+
+  getResourceKey(resourceType, resourceId, key) {
+    logger.debug(`Getting resource key: ${key} for ${resourceType}/${resourceId}`);
+    const resourceFolderName = this.getResourceFolderName(resourceType, resourceId);
+    const resourceKey = `${resourceFolderName}/${key}`;
+    return etcd.get(resourceKey).string();
+  }
+
+  getResourceState(resourceType, resourceId) {
+    return this.getResourceKey(resourceType, resourceId, CONST.RESOURCE_KEYS.STATE);
+  }
+
+  registerWatcher(key, callback, isRecursive) {
+    if (isRecursive === true) {
+      logger.debug(`Registering recursive watcher on prefix: ${key}`);
+      return etcd.watch()
+        .prefix(key)
+        .create()
+        .then(watcher => {
+          return watcher
+            .on('put', callback);
+        });
+    } else {
+      logger.debug(`Registering watcher on key: ${key}`);
+      return etcd.watch()
+        .key(key)
+        .create()
+        .then(watcher => {
+          return watcher
+            .on('put', callback);
+        });
+    }
+  }
+
+  annotateResource(resourceType, resourceId, annotationName, operationType, opId, val) {
+    const annotationFolderName = this.getAnnotationFolderName(resourceType, resourceId, annotationName, operationType, opId);
+    const optionKey = `${annotationFolderName}/${CONST.ANNOTATION_KEYS.OPTONS}`;
+    logger.debug(`Annotating Resource ${resourceType}/${resourceId} for annotation: ${annotationName}`);
+    return etcd.put(optionKey).value(val)
+      .then(() => {
+        const statusKey = `${annotationFolderName}/${CONST.ANNOTATION_KEYS.STATE}`;
+        return etcd.put(statusKey).value(CONST.RESOURCE_STATE.IN_QUEUE);
+      });
+  }
+
+  updateAnnotationState(resourceType, resourceId, annotationName, operationType, opId, stateValue) {
+    return this.updateAnnotationKey(resourceType, resourceId, annotationName, operationType, opId, CONST.ANNOTATION_KEYS.STATE, stateValue);
+  }
+
+  updateAnnotationKey(resourceType, resourceId, annotationName, operationType, opId, key, value) {
```

Comment:
Can this function be directly called from outside of this class? If yes, then we need validation on key, if not, is there a way to make it private?